src/AsyncBridge/AsyncCompatLibExtensions.cs

```diff
@@ -168,29 +168,28 @@ public static void CancelAfter(this CancellationTokenSource cancelSource, TimeSp
         if (cancelSource.IsCancellationRequested)
             return;
 
-        Timer MyTimer = null;
+        Timer myTimer = null;
 
         // CTS claims it's thread-safe for all methods, so we should be the same.
         // If someone calls CancelAfter concurrently for the same CTS, we won't explode or cause a memory leak
-        do
+        while (!_cancelTimers.TryGetTimer(cancelSource, out myTimer))
         {
-            if (_cancelTimers.TryGetTimer(cancelSource, out var CurrentTimer))
-            {
-                if (MyTimer != null)
-                    MyTimer.Dispose();
-
-                MyTimer = CurrentTimer;
-                break; // Reset the existing timer to the new delay
-            }
-
             // An active timer doesn't hold a strong reference to itself, so adding the CTS as state won't hold it alive
-            MyTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
+            myTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
 
-        } while (!_cancelTimers.TryAddTimer(cancelSource, MyTimer));
+            if (_cancelTimers.TryAddTimer(cancelSource, myTimer))
+                break;
+
+            // TryAddTimer can only fail if a Timer was created concurrently. Since we never remove them, TryGetTimer is guaranteed to succeed
```

Comment:
The "we never remove them" doesn't sound quite right. A previous CancelAfter may occur, doing `TryRemoveTimer` on a different thread at this point so that `TryGetTimer` does not succeed. Maybe we should check `cancelSource.IsCancellationRequested` right before creating the new timer, just to save resources if it was concurrently canceled?

Reply:
You are correct :)
Thread 1: IsCancellationRequested returns false
ThreadPool: Timer elapses, callback is queued and triggered, calls TryRemoveTimer. CTS gets cancelled
Thread 1: TryGetTimer returns false. TryAddTimer succeeds, we set up our Timer

Technically at this point there's a Timer that shouldn't exist, and sure - adding another IsCancellationRequested call will fix that. But equally, the sequence could be:

Thread 1: IsCancellationRequested returns false
ThreadPool: Timer elapses, callback is queued and triggered, calls TryRemoveTimer
Thread 1: TryGetTimer returns false. TryAddTimer succeeds, we set up our Timer
ThreadPool:  CTS gets cancelled

Now the extra check does nothing - we end up with the same situation. We could throw locks around everything, but that's pretty heavy for something that won't happen very often. And in the worst case, we'll just have another Timer sitting around waiting to elapse. And it'll only exist as long as someone holds a reference to the CTS - as soon as it's available for GC, the Timer will be too.

Personally, I'd prefer to avoid adding extra checks for rare cases, as long as those rare cases don't result in a memory leak (which this won't for long - it'll be GC'ed along with the CTS) or state corruption. I'll just fix the comment up.